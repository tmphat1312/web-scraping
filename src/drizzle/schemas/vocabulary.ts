import { integer, sqliteTable, text } from 'drizzle-orm/sqlite-core';

export const vocabulary = sqliteTable('vocabulary', {
  id: integer('id').primaryKey({ autoIncrement: true }),
  level: integer('level').notNull(),
  character: text('character').notNull(),
  primary_meaning: text('primary_meaning').notNull(),
  alternative_meanings: text('alternative_meanings'),
  word_types: text('word_types').notNull(),
  meaning_explanation: text('meaning_explanation').notNull(),
  reading_explanation: text('reading_explanation'),
});
