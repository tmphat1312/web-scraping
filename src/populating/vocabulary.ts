import { fromURL } from 'cheerio';

scrapeVocabulary('大人').then((data) => console.log(data.context_sentences));

export async function scrapeVocabulary(slug: string) {
  let $ = await fromURL(`https://www.wanikani.com/vocabulary/${slug}`);
  return {
    vocabulary: $.extract({
      level: {
        selector: `[class$="icon--level"]`,
        value: (el) => Number($(el).text()),
      },
      character: `[class$="icon--vocabulary"]`,
      primary_meaning: `.subject-section__meanings-items:first`,
      alternative_meanings: {
        selector: `.subject-section__meanings:eq(1)`,
        value: (el) => {
          let title = $(el).find('.subject-section__meanings-title').text();
          let item = $(el).find('.subject-section__meanings-items').text();
          return title.includes('Alternative') ? item : undefined;
        },
      },
      word_types: {
        selector: `.subject-section__meanings:last`,
        value: (el) => {
          let title = $(el).find('.subject-section__meanings-title').text();
          let item = $(el).find('.subject-section__meanings-items').text();
          return title.includes('Word') ? item : undefined;
        },
      },
      meaning_explanation: [
        {
          selector: `#section-meaning .subject-section__text`,
          value: (el) => $(el).html(),
        },
      ],
      reading_explanation: [
        {
          selector: `#section-reading .subject-section__text`,
          value: (el) => $(el).html(),
        },
      ],
    }),
    vocabulary_readings: $.extract({
      character: `[class$="icon--vocabulary"]`,
      reading_with_audios: [
        {
          selector: `.reading-with-audio`,
          value: (el) => {
            return $(el).extract({
              reading: `[lang="ja"]`,
              audios: [
                {
                  selector: `li`,
                  value: {
                    sources: {
                      selector: `source`,
                      value: (el) => $(el).attr('src'),
                    },
                    actor_name: `.reading-with-audio__voice-actor-name`,
                  },
                },
              ],
            });
          },
        },
      ],
    }),
    context_sentences: $.extract({
      character: `[class$="icon--vocabulary"]`,
      sentences: [
        {
          selector: `#section-context .subject-section__text`,
          value: (el) => {
            return $(el).extract({
              jp_sentence: `p:first`,
              en_sentence: `p:last`,
            });
          },
        },
      ],
    }),
    patterns_of_use_context: $.extract({
      character: `[class$="icon--vocabulary"]`,
      patterns_of_use: [`.subject-collocations__pattern-name`],
      common_word_combinations: [
        {
          selector: `.context-sentences`,
          value: (el) => {
            return $(el).extract({
              jp_sentence: `p:first`,
              en_sentence: `p:last`,
            });
          },
        },
      ],
    }),
  };
}
