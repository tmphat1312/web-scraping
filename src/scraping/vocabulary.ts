import { fromURL } from 'cheerio';

export async function scrapeVocabulary(slug: string) {
  let $ = await fromURL(`https://www.wanikani.com/vocabulary/${slug}`);
  return $.extract({
    level: {
      selector: `[class$="icon--level"]`,
      value: (el) => Number($(el).text()),
    },
    character: `[class$="icon--vocabulary"]`,
    primaryMeaning: `.subject-section__meanings-items:first`,
    alternativeMeanings: {
      selector: `.subject-section__meanings:eq(1)`,
      value: (el) => {
        let title = $(el).find('.subject-section__meanings-title').text();
        let item = $(el)
          .find('.subject-section__meanings-items')
          .text()
          .split(', ');
        return title.includes('Alternative') ? item : undefined;
      },
    },
    wordTypes: {
      selector: `.subject-section__meanings:last`,
      value: (el) => {
        let title = $(el).find('.subject-section__meanings-title').text();
        let item = $(el)
          .find('.subject-section__meanings-items')
          .text()
          .split(', ');
        return title.includes('Word') ? item : undefined;
      },
    },
    meaningExplanation: [
      {
        selector: `#section-meaning .subject-section__text`,
        value: (el) => $(el).html(),
      },
    ],
    readingExplanation: [
      {
        selector: `#section-reading .subject-section__text`,
        value: (el) => $(el).html(),
      },
    ],
    contextSentences: [
      {
        selector: `#section-context .subject-section__text`,
        value: (el) => {
          return $(el).extract({
            jaSentence: `p:first`,
            enSentence: `p:last`,
          });
        },
      },
    ],
    readings: [
      {
        selector: `.reading-with-audio__audio-item`,
        value: (el) => {
          let readingEl = $(el).closest('.reading-with-audio');
          return {
            source: $(el).find('source').attr('src')!,
            actor: $(el).find(`[class$="voice-actor-name"]`).text(),
            reading: $(readingEl).find(`[lang="ja"]`).text(),
          };
        },
      },
    ],
    commonUsagePatterns: [
      {
        selector: `.subject-collocations__pattern-name`,
        value: (el) => {
          let targetID = $(el).attr('aria-controls');
          return {
            pattern: $(el).text(),
            sentences: $(`#${targetID}`).extract({
              sentences: [
                {
                  selector: `.context-sentences`,
                  value: (el) => {
                    return $(el).extract({
                      jaSentence: `p:first`,
                      enSentence: `p:last`,
                    });
                  },
                },
              ],
            }).sentences,
          };
        },
      },
    ],
  });
}
