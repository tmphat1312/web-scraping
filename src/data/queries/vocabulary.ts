import { InferSelectModel, eq } from 'drizzle-orm';

import { db } from '../client';
import { vocabulary } from '../schemas/vocabulary';

export type Vocabulary = InferSelectModel<typeof vocabulary>;

export async function getVocabularyByCharacter(character: string) {
  return db
    .select({ id: vocabulary.id })
    .from(vocabulary)
    .where(eq(vocabulary.character, character));
}
