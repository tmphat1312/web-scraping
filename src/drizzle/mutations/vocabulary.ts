import { InferInsertModel } from 'drizzle-orm';

import { db } from '../client';
import { vocabulary } from '../schemas/vocabulary';

export type Vocabulary = InferInsertModel<typeof vocabulary>;

export async function insertVocabulary(data: Vocabulary[]) {
  return db
    .insert(vocabulary)
    .values(data)
    .returning({
      id: vocabulary.id,
    })
    .onConflictDoNothing();
}
