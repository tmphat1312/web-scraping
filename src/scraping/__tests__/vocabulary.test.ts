import { scrapeVocabulary } from '../vocabulary';

describe('scrape vocabulary', () => {
  it('should be the same vocabulary', async () => {
    let vocabulary = await scrapeVocabulary('æ—¥');
    expect(vocabulary).toMatchSnapshot();
  });
});
