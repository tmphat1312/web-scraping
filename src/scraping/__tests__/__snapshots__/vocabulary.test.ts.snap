// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scrape vocabulary > should be the same vocabulary 1`] = `
{
  "alternativeMeanings": [
    "Day",
  ],
  "character": "日",
  "commonUsagePatterns": [
    {
      "pattern": "お日〜",
      "sentences": [
        {
          "enSentence": "the sun",
          "jpSentence": "お日さま",
        },
        {
          "enSentence": "the sun",
          "jpSentence": "お日さん",
        },
      ],
    },
    {
      "pattern": "〜日",
      "sentences": [
        {
          "enSentence": "this day",
          "jpSentence": "この日",
        },
        {
          "enSentence": "that day",
          "jpSentence": "その日",
        },
        {
          "enSentence": "one day, a certain day",
          "jpSentence": "ある日",
        },
      ],
    },
    {
      "pattern": "日が〜",
      "sentences": [
        {
          "enSentence": "to get dark",
          "jpSentence": "日がくれる",
        },
        {
          "enSentence": "the sun rises",
          "jpSentence": "日がのぼる",
        },
        {
          "enSentence": "the sun sets",
          "jpSentence": "日がしずむ",
        },
      ],
    },
  ],
  "contextSentences": [
    {
      "enSentence": "The sun will come out soon.",
      "jaSentence": "そろそろ日が出ますよ。",
    },
    {
      "enSentence": "The date the puppy comes is the 3rd.",
      "jaSentence": "子犬がくる日は三日です。",
    },
    {
      "enSentence": "I lost my passport on the day I came to Japan.",
      "jaSentence": "日本にきた日にパスポートをなくした。",
    },
  ],
  "level": 2,
  "meaningExplanation": [
    "The kanji and the word are exactly the same. That means they share meanings as well. Take note that this word means both "sun" and "day," though. Why? because the sun takes a day to go around and come back!",
  ],
  "primaryMeaning": "Sun",
  "readingExplanation": [
    "Since this word is made up of a single kanji, it uses the kun'yomi reading. Here's a mnemonic to help you to remember it: ",
    "What do you feel from the sun? You feel the <mark title="Reading" class="reading-highlight">hea</mark>t (<span lang="ja">ひ</span>).",
  ],
  "readings": [
    {
      "actorName": "Kyoko",
      "reading": "ひ",
      "source": "https://files.wanikani.com/wlzsrjdvwo8hcav432mh97ouxfxk",
    },
    {
      "actorName": "Kenichi",
      "reading": "ひ",
      "source": "https://files.wanikani.com/v7zxwkl4apkovzc1xfky6l248i8h",
    },
  ],
  "wordTypes": [
    "noun",
  ],
}
`;
